### PR TITLE
Add JointPositionGoalController

### PR DESCRIPTION
**Pull request by wxmerkt**
Adds a controller that expects a JOINT_POSITION_GOAL message and executes it adhering to conservative, definable joint position constraints (using ROSControl's preferred JointLimitsInterface) as well as a joint velocity constraint. Example launch file for controlling the neck included.

Tested in simulation with Gazebo:
[JointPositionGoalController](https://www.youtube.com/watch?v=MtgpuEE7rHE)
https://www.youtube.com/watch?v=MtgpuEE7rHE

### CODE REVIEW

**Comment by mauricefallon** [comment]
can you remove the include directory and place all the code under src?

File: config/NeckControl.yaml [hunk 1]
```diff
@@ -1,6 +1,21 @@
-LCM2ROSControl:
-    type: valkyrie_translator/LCM2ROSControl
+JointPositionGoalController:
+    type: valkyrie_translator/JointPositionGoalController
+    publish_est_robot_state: true
     joints:
       - upperNeckPitch
       - lowerNeckPitch
       - neckYaw
+    joint_velocity_limit: 15 # deg/s
```
Comment (mauricefallon) [inline config/NeckControl.yaml]: mixture of deg/s and radians. change this to rad/s

File: config/NeckControl.yaml [hunk 2]
```diff
@@ -1,6 +1,21 @@
...
+    joint_limits:
+      upperNeckPitch:
```
Comment (mauricefallon) [inline config/NeckControl.yaml:10]: i don't like hard coding the limits in this xml. i'll accept it for now

File: package.xml [hunk 3]
```diff
@@ -2,7 +2,7 @@
 <package format="2">
   <name>valkyrie_translator</name>
   <version>0.0.1</version>
-  <description>Package providing translators between Drake controller and Valkyrie ROS Control system.</description>
+  <description>Package providing translators between Drake controller and Valkyrie ROS Control system as LCM-exposed controllers.</description>
```
Comment (mauricefallon) [inline package.xml]: its not a translator its a mid level controller

File: src/JointPositionGoalController.cpp [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+#include <joint_limits_interface/joint_limits.h>
+#include <joint_limits_interface/joint_limits_rosparam.h>
+
+#include <lcm/lcm-cpp.hpp>
+#include "lcmtypes/bot_core/joint_state_t.hpp"
+#include "lcmtypes/bot_core/robot_state_t.hpp"
+
+inline double clamp(double x, double lower, double upper) {
+    return x < lower ? lower : (x > upper ? upper : x);  // TODO: move to std::min std::max
```
Comment (mauricefallon) [inline src/JointPositionGoalController.cpp]: move to std:min or std:max before merging

File: src/JointPositionGoalController.cpp [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        bool use_joint_selection = true;
+        if (n_joints_ == 0)
+            use_joint_selection = false;
+
+        // Retrieve maximum joint velocities in degrees per second and convert to radians per tick
+        if (!controller_nh.getParam("joint_velocity_limit", max_joint_velocity_)) {
+            ROS_WARN("Cannot retrieve desired maximum joint velocity from param server, defaulting to 10deg/s");
+            max_joint_velocity_ = 10;  // Default to 10deg/s if not set
+        }
+        max_joint_velocity_rad_ = (max_joint_velocity_ * M_PI / 180);
```
Comment (mauricefallon) [inline src/JointPositionGoalController.cpp]: modify to be only radians

divide by 180.0

File: src/JointPositionGoalController.cpp [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    void JointPositionGoalController::starting(const ros::Time &time) {
+        last_update_ = time;
+    }
+
+    void JointPositionGoalController::update(const ros::Time &time, const ros::Duration &period) {
+        handler_->update();
+        lcm_->handleTimeout(0);
+
+        double dt = (time - last_update_).toSec();
+        int64_t utime = (int64_t) (time.toSec() * 1000000.);
```
Comment (mauricefallon) [inline src/JointPositionGoalController.cpp]: Try to use 1E6

File: src/JointPositionGoalController.cpp [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+
+        size_t number_of_joint_interfaces = positionJointHandles_.size();
+
```
Comment (mauricefallon) [inline src/JointPositionGoalController.cpp]: Can we remove these diagnostics publishers (preferred) or share them with the other LCM2ROS controller.

They should be functionalised that than dropped into the middle of a long function

File: src/JointPositionGoalController.cpp [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        lcm_state_msg.pose.rotation.y = 0.0;
+        lcm_state_msg.pose.rotation.z = 0.0;
+        lcm_state_msg.twist.linear_velocity.x = 0.0;
+        lcm_state_msg.twist.linear_velocity.y = 0.0;
+        lcm_state_msg.twist.linear_velocity.z = 0.0;
+        lcm_state_msg.twist.angular_velocity.x = 0.0;
+        lcm_state_msg.twist.angular_velocity.y = 0.0;
+        lcm_state_msg.twist.angular_velocity.z = 0.0;
+
+        // Iterate over all position-controlled joints
```
Comment (mauricefallon) [inline src/JointPositionGoalController.cpp:209]: Same here. Make a function called publishRobotStateToLCM

See how the Atlas driver is split up into different files

File: src/JointPositionGoalController.cpp [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+            std::cerr << "ERROR: handler lcm is not good()" << std::endl;
+        }
+        lcm_->subscribe("JOINT_POSITION_GOAL", &JointPositionGoalController_LCMHandler::jointPositionGoalHandler, this);
+    }
+
+    JointPositionGoalController_LCMHandler::~JointPositionGoalController_LCMHandler() { }
+
+    void JointPositionGoalController_LCMHandler::jointPositionGoalHandler(const lcm::ReceiveBuffer *rbuf,
+                                                                          const std::string &channel,
+                                                                          const bot_core::robot_state_t *msg) {
```
Comment (mauricefallon) [inline src/JointPositionGoalController.cpp:346]: i don't think this is the message we want to use in this API, but lets change it later

**Comment by wxmerkt** [comment]
Thanks for the review. I've addressed all comments.

**Comment by mauricefallon** [comment]
Some notes to take forward:
- change LCM2ROSControl library name to better match conventions
- reconsider the source for joint limits
- decide on message I/O:
  - type might be joint_angles
  - message name (now: JOINT_POSITION_GOAL) might be better as DESIRED_NECK_ORIENTATION (when used specifically). Also want to use the same neck message from Atlas. 
- Look at how this meshes with interface to IHMC in simulation (where we control the neck and forearms)

**Pull request merged**